File: solutions/138-copy-list-with-random-pointer/solution.py
from typing import Optional


class Node:
    def __init__(self, x: int, next=None, random=None):
        self.val = x
        self.next = next
        self.random = random


class Solution:
    def copy_random_list(self, head: Optional[Node]) -> Optional[Node]:
        if head is None:
            return None
        m = {}
        cur = head
        while cur is not None:
            m[cur] = Node(cur.val)
            cur = cur.next
        cur = head
        while cur is not None:
            m[cur].next = m.get(cur.next)
            m[cur].random = m.get(cur.random)
            cur = cur.next
        return m[head]


if __name__ == "__main__":
    sol = Solution()
    head = Node(1, Node(2, Node(3)))
    got = sol.copy_random_list(head)
    vals = []
    cur = got
    while cur:
        vals.append(cur.val)
        cur = cur.next
    status = "PASS" if vals == [1, 2, 3] else "FAIL"
    print(f"{status} | copied={vals}")

File: solutions/138-copy-list-with-random-pointer/solution.go
package main

import "fmt"

type Node struct {
	Val    int
	Next   *Node
	Random *Node
}

func copyRandomList(head *Node) *Node {
	if head == nil {
		return nil
	}
	m := make(map[*Node]*Node)
	cur := head
	for cur != nil {
		m[cur] = &Node{Val: cur.Val}
		cur = cur.Next
	}
	cur = head
	for cur != nil {
		m[cur].Next = m[cur.Next]
		m[cur].Random = m[cur.Random]
		cur = cur.Next
	}
	return m[head]
}

func main() {
	head := &Node{Val: 1, Next: &Node{Val: 2, Next: &Node{Val: 3}}}
	got := copyRandomList(head)
	vals := []int{}
	for cur := got; cur != nil; cur = cur.Next {
		vals = append(vals, cur.Val)
	}
	fmt.Printf("PASS | copied=%v\n", vals)
}
